### PR TITLE
Math tutorial

### PR DESCRIPTION
**Pull request by bocchino**
| | |
|:---|:---|
|**_Originating Project/Creator_**| F Prime|
|**_Affected Component_**| Math tutorial |
|**_Affected Architectures(s)_**| Ref |
|**_Related Issue(s)_**| N/A |
|**_Has Unit Tests (y/n)_**| y |
|**_Builds Without Errors (y/n)_**| y |
|**_Unit Tests Pass (y/n)_**| y |
|**_Documentation Included (y/n)_**| y |

---
## Change Description

This PR updates the tutorial at docs/MathComponent to use FPP instead of XML as the source language. It also makes two incidental changes that are necessary for the tutorial to work as described:

1. Revise the component implementation template autocoder to use the name _Component_ instead of _Component_ `ComponentImpl`. This change conforms to the latest F Prime conventions.
2. Restore console logging in the Ref application by implementing a subclass of `Fw::Logger` that prints to the console. It appears that Ref console logging was dropped at some point. I think it is useful for the Ref app to show how to use the logger interface.

## Rationale

This PR includes essential documentation for release 3.0.

## Testing/Review Recommendations

I made some changes to the content of the tutorial. Specifically, I simplified and revised the use of types, I simplified the command interface, and I refactored the unit tests to remove code duplication between tests. I also removed image files and sequence files that were out of date and that were not mentioned in the tutorial. You may want to review these changes.

I recommend that we go with this version, which I believe suffices for release 3.0. If we want to add more complexity, we can add another tutorial or we can update this tutorial later. I believe the current tutorial covers all the essential features of FPP modeling, building, and testing.

I have delivered the revised tutorial in AsciiDoc. I believe it is a much better choice than Markdown for moderately complex documentation like this. We have an agenda item to discuss the merits of using AsciiDoc vs. sticking with Markdown.

## Future Work

* Once we have `fpp-check -u` integrated with the build system, update this tutorial to explain how to use it.
* Update the other tutorials.

### CODE REVIEW

File: Ref/Top/Main.cpp [hunk 1]
```diff
@@ -11,7 +12,33 @@ void print_usage(const char* app) {
 #include <signal.h>
 #include <cstdio>
 
+class Logger :
```
Comment (LeStarch) [inline Ref/Top/Main.cpp:15]: This is typically done by the following:

```
#include <Os/Log.hpp>

Os::Log logger;
```

This will use the Os::Log implementation as the logger.  No need to define a custom logger here.  My guess is this code was done in the files replace by the autocoder output.

File: docs/Tutorials/MathComponent/sequences/math_sequence.seq [hunk 2]
```diff
@@ -1,9 +0,0 @@
-; Run this sequence to replicate the MathComponent tutorial
```
Comment (LeStarch) [inline docs/Tutorials/MathComponent/sequences/math_sequence.seq:1]: Why was this deleted?  Are we not covering sequences anymore?  Also, it looks out of date.

**Review by LeStarch — COMMENTED** [review]
I see a lot of utility scripts and some JSON checked in.  Is this needed?

What are the ramifications of changing the component autocoded name in the template files?  Is this going to break `fprime-util new` or other tooling?

File: Ref/Top/Main.cpp (hunk 1 above)
Comment (bocchino) [reply Ref/Top/Main.cpp:15]: Good point. I will fix it.

File: docs/Tutorials/MathComponent/sequences/math_sequence.seq (hunk 2 above)
Comment (bocchino) [reply docs/Tutorials/MathComponent/sequences/math_sequence.seq:1]: I didn't see anything in the previous version of the tutorial that discussed this sequence. It was just there. I see two options going forward:

1. Revise the sequence to conform to the new tutorial and add a section of the tutorial discussing the sequence.
2. Delete the sequence.

I went with option (2) because sequencing is covered elsewhere and the tutorial already seems long enough. Thoughts?

**Comment by bocchino** [comment]
> I see a lot of utility scripts and some JSON checked in. Is this needed?

The utility scripts in `adoc/` are to avoid hand-copying the reference application into the AsciiDoc document. For maintainability, I recommend that we leave them in.

In `json/`, `json.txt` is the F Prime Layout form of the example diagram. I recommend keeping it for reference. We could delete it because it is auto-generated from `top.txt`.

**Comment by bocchino** [comment]
> What are the ramifications of changing the component autocoded name in the template files? Is this going to break fprime-util new or other tooling?

I did a quick test. `fprime-util new` doesn't seem to care about the class name in the implementation. It does assume that the implementation file is named FooComponent.{hpp,cpp} or FooComponentImpl.{hpp,cpp). So to use release 3.0 naming convention (also expected by FPP), you need to do a bit of renaming after invoking `fprime-util new`. This seems like an `fprime-util new` issue, not a math tutorial issue.

I'm not aware of any other tool that will break. Also, if some tool does break, we should fix the tool, since this naming scheme is the release 3.0 convention.

File: docs/Tutorials/MathComponent/sequences/math_sequence.seq (hunk 2 above)
Comment (LeStarch) [reply docs/Tutorials/MathComponent/sequences/math_sequence.seq:1]: Delete the sequence.  It doesn't even follow the correct naming any more.

**Comment by LeStarch** [comment]
@bocchino it is an fprime-util new issue.  Just wanted to confirm that fprime-util new did not crash or something.  We do need to rework it to put in FPP files and better naming conventions.
